api/v1alpha1/servicedefaults_types.go

```diff
@@ -52,25 +58,44 @@ func init() {
 }
 
 // ToConsul converts the entry into it's Consul equivalent struct.
-func (s *ServiceDefaults) ToConsul() *capi.ServiceConfigEntry {
+func (in *ServiceDefaults) ToConsul() *capi.ServiceConfigEntry {
 	return &capi.ServiceConfigEntry{
 		Kind: capi.ServiceDefaults,
-		Name: s.Name,
-		//Namespace: s.Namespace, // todo: don't set this unless enterprise
-		Protocol:    s.Spec.Protocol,
-		MeshGateway: s.Spec.MeshGateway.toConsul(),
-		Expose:      s.Spec.Expose.toConsul(),
-		ExternalSNI: s.Spec.ExternalSNI,
+		Name: in.Name,
+		//Namespace: in.Namespace, // todo: don't set this unless enterprise
+		Protocol:    in.Spec.Protocol,
+		MeshGateway: in.Spec.MeshGateway.toConsul(),
+		Expose:      in.Spec.Expose.toConsul(),
+		ExternalSNI: in.Spec.ExternalSNI,
 	}
 }
 
 // MatchesConsul returns true if entry has the same config as this struct.
-func (s *ServiceDefaults) MatchesConsul(entry *capi.ServiceConfigEntry) bool {
-	return s.Name == entry.GetName() &&
-		s.Spec.Protocol == entry.Protocol &&
-		s.Spec.MeshGateway.Mode == string(entry.MeshGateway.Mode) &&
-		s.Spec.Expose.matches(entry.Expose) &&
-		s.Spec.ExternalSNI == entry.ExternalSNI
+func (in *ServiceDefaults) MatchesConsul(entry *capi.ServiceConfigEntry) bool {
+	return in.Name == entry.GetName() &&
+		in.Spec.Protocol == entry.Protocol &&
+		in.Spec.MeshGateway.Mode == string(entry.MeshGateway.Mode) &&
+		in.Spec.Expose.matches(entry.Expose) &&
+		in.Spec.ExternalSNI == entry.ExternalSNI
+}
+
+// Validate validates the fields provided in the spec of the ServiceDefaults and
+// returns an error which lists all invalid fields in the resource spec.
+func (in *ServiceDefaults) Validate() error {
+	var allErrs field.ErrorList
```

Comment:
What about validating `protocol`? I thought only tcp, http, http2 were supported?

Reply:
I used [this page](https://www.consul.io/docs/agent/config-entries/service-defaults) to determine the valid values which did not specify that subset. Where did you find your list? Maybe there are some others I might have missed.

Reply:
Gotcha, looking at the consul code there is no further validation (afaict) so nvm.